Update CI Page (#260)
NUbots/NUbook: src/book/02-system/01-foundations/03-continuous-integration.mdx

  - Josephus Paye II (@JosephusPaye)
  - Joel Wong (@wongjoel)
  - Yosiah de Koeyer (@Y0sh1dk)
  - Kip Hamiltons (@KipHamiltons)
---

## Overview

NUbots uses [GitHub Actions](https://github.com/features/actions) for its continuous integration pipelines. The pipelines are triggered automatically when pull requests are made, updated, or merged. The pipelines build and test the code on GitHub's cloud, with logs and results available on the GitHub website, associated with the relevant pull request.

## Branch Workflow

**This section assumes basic knowledge of Git. If you need a git refresher, see the [Git guide](/guides/general/git).**

This CI process works in combination with the [Git Feature Branch workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) we use for development:

1. Developers branch off `main`, creating a new branch with a name of the form `<lastname>/<purpose-of-branch>`. For example:

   ```bash
   # Make sure main is up to date before creating your branch
   git fetch
   git checkout main
   git pull

   # Create and check out the branch off of main
   git checkout -b biddulph/ball-detector
   ```

> 2. Once development on the branch is complete, or reaches a point where it can be reviewed, the developer makes a pull request (PR) on GitHub. Creating the pull request triggers a GitHub action runners to check the code. GitHub automatically creates separate build runners for different parts of the PR (build, check code format, test, etc).

Author: I might reword to "GitHub action agents" to keep it consistent with the other wording

> 3. Each agent completes a set of tasks specific in a pipeline file, found inside the [`.github/workflows`](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows) directory of the codebase. These steps usually involve cloning the codebase, checking out the relevant commit, pulling down the docker image (rebuilding it if it has changed), then checking some aspect of the code.

Reviewer: This link should reference the main branch
```suggestion
3. Each agent completes a set of tasks specified in a workflow file, found inside the [`.github/workflows`](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows) directory of the codebase. These steps usually involve cloning the codebase, checking out the relevant commit, pulling down the docker image (rebuilding it if it has changed), then checking some aspect of the code.
```

4. GitHub logs the results of each job and displays the results on the webpage for the pull request. The logs are viewable in the 'actions' tab of the PR.

> 5. Once the build passes, the PR is reviewed by other members of the team. Once the PR has been approved, it can be merged into `main`.

Reviewer: ```suggestion
5. Once the build passes, the PR is reviewed by other members of the team. Once the PR has been approved and all required checks pass, it can be merged into `main`.
```

6. Merging the PR into `main` then triggers another CI build. This time, in addition to building the code, a separate job is created to rebuild the main docker images and push them to DockerHub. This is the main way that our docker images are kept up to date.

> <TabbedImages>

Reviewer 2: Can't comment on the line directly but the image caption and file name can be updated to `main` instead of `master`.

## The Build Pipeline

The GitHub Actions are configured in [workflow files](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows) in the NUbots repo. To add, remove, or change build jobs, we edit the relevant workflow file. The [GitHub Actions documentation](https://docs.github.com/en/actions) should be your first stop when you want to make changes.

## Monitoring and Troubleshooting Builds

GitHub Actions have live-updating logs of each build, which you can use to monitor the build progress or troubleshoot errors when a build fails.

Failed builds and pipeline steps can be diagnosed by viewing their logs.

> To view them, scroll to the build checks at the bottom of the PR page and click the **Details** link on the `buildkite/nubots/pr` check.

Author: need to update this too

## The Build Pipeline

> GitHub Actions are configured in [workflow files](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions), written in YAML. To add, remove, or change build jobs, we edit the relevant [workflow file in the NUbots repo](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows). The [GitHub Actions documentation](https://docs.github.com/en/actions) should be your first stop when you want to make changes.

Reviewer: This link also need to reference the main branch

Author: permalinks are the good stuff

Reviewer: You want links to the commits rather than to main itself?

Reviewer 3: +1 for permalinks but I'll grab a more recent commit at least. there's pros/cons to both but an out of date permalink is low risk low impact compare to a broken link imo.

Author: > You want links to the commits rather than to main itself?

If the path changes on main and you linked main, you have to start archaeology immediately to get any idea of the file contents. If you have a permalink, you can check main with a click, but you're guaranteed to get something useful even if it changed on main.

Reviewer: Sure, if you say so

Author: We take that as a W

> GitHub Actions are configured in [workflow files](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions), written in YAML. To add, remove, or change build jobs, we edit the relevant [workflow file in the NUbots repo](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows). The [GitHub Actions documentation](https://docs.github.com/en/actions) should be your first stop when you want to make changes.

Reviewer 2: Can't comment on the line directly but the image caption and file name can be updated to `main` instead of `master`.

## Workflow Outline

**Note that the source of truth for what a workflow actually does is the given workflow file. It defines precisely what happens.**

When a pull request is created or its branch is modified, checks are run on the code to make sure that it builds and maintains standards of correctness. First, an action checks which parts of the codebase have been modified, which informs which checks should be run. For instance, changes to C++ code require the C++ code to be built, and its formatting checked.

> Next, a docker image tag is built and pushed to DockerHub for the PR. This is cached from the PR image, if it already exists, and the `main` branch image. The caching means that only what needs to be rebuilt in the image is rebuilt, to the point that if the image hasn't changed, the PR image tag simply acts as an alias for the image it cached from.

Author: I agree this is definitely awkward wording. I think your suggestion works. There is the third case, that neither cache is valid for the changed dockerfile, in which case the thing is just built from the layer which diverges, but I don't think that's really an essential detail _we_ need to capture in our docs.

Finally, all other checks are run conditionally, based on whether the files which those checks apply to have changed. These checks are run inside the Docker container associated with the PR image.

The results from each of these steps are displayed on the pull request page on GitHub (see below).

Each of these steps happens when a pull request is merged into the `main` branch. The results are displayed on the main page of the repository though, rather than the pull request page.

> Additionally, when pull requests are merged into main, docker images are built, tagged, and pushed to be the new standard images we use - for example, the `generic` image used for most of our development. This actions workflow is the primary way that those standard images are updated. They can be updated manually by pushing a local image, but this should only be done in exceptional circumstances.

Reviewer 4: We should document how to do this at some point but I think it's out of scope for this page

Author: Should we document that? I think it's as simple as `docker tag nubots/nubots:generic && docker push nubots/nubots:generic`, which is what you would expect, right? Is that something we need to cover in our docs?

In saying that, I feel like perhaps I know too much. Maybe I have lost touch with what nubook should have, since I am familiar with these tools now.

Reviewer 4: Personally I didn't know how to do that although I would probably have been able to figure it out by looking though the docker documentation. I think it's a classic case of at what point does it change from documentation to tutorials. Probs fine to leave out tbh

> Additionally, when pull requests are merged into main, docker images are built, tagged, and pushed to be the new standard images we use - for example, the `generic` image used for most of our development. This actions workflow is the primary way that those standard images are updated. They can be updated manually by pushing a local image, but this should only be done in exceptional circumstances.

Reviewer: I think either remove this last sentence or move it into a "details" section (you can then include basic instructions and a link to relevant external documentation if you use a details section). This is not an action that should be run (even if you manage to push broken/corrupted images to DockerHub you should still be able to fix them via a new PR) and the vast majority of people on the team don't even need to know that this is a possibility.
```suggestion
Additionally, when pull requests are merged into main, docker images are built, tagged, and pushed to be the new standard images we use - for example, the `generic` image used for most of our development. This actions workflow is the primary way that those standard images are updated.
```
